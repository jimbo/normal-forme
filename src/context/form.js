import { createContext, useCallback, useContext } from "react"

import ScopeContext from "../context/scope"
import { getDeepValue, join } from "../util/deep"

const FormContext = createContext()
const { Consumer, Provider } = FormContext

export default FormContext
export { Consumer as FormConsumer }
export { Provider as FormProvider }

export const useFormContext = field => {
  const scope = useContext(ScopeContext)
  const scopedField = join(scope, field)
  const [state, dispatch] = useContext(FormContext)
  const value = getDeepValue(state.valueMap, scopedField)

  const setValue = useCallback(
    nextValue =>
      dispatch({
        type: "set value",
        payload: { field: scopedField, nextValue },
      }),
    [dispatch, scopedField]
  )

  const transformValue = useCallback(
    transformValue =>
      dispatch({
        type: "transform value",
        payload: { field: scopedField, transformValue },
      }),
    [dispatch, scopedField]
  )

  return { setValue, transformValue, value }
}
