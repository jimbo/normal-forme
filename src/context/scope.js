import { createContext } from "react"

const ScopeContext = createContext("")
const { Consumer, Provider } = ScopeContext

export default ScopeContext
export { Consumer as ScopeConsumer }
export { Provider as ScopeProvider }
