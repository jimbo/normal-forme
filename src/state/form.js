import { useReducer } from "react"

import { getDeepValue, setDeepValue } from "../util/deep"

const getInitialState = (initialValueMap = new Map()) => ({
  valueMap: initialValueMap,
})

export const reducer = (state = getInitialState(), { payload, type } = {}) => {
  switch (type) {
    case "set value": {
      const { valueMap } = state
      const { field, nextValue } = payload
      const nextValueMap = setDeepValue(valueMap, field, nextValue)

      return { valueMap: nextValueMap }
    }
    case "transform value": {
      const { valueMap } = state
      const { field, transformValue } = payload
      const currentValue = getDeepValue(valueMap, field)
      const nextValue = transformValue(currentValue)
      const nextValueMap = setDeepValue(valueMap, field, nextValue)

      return { valueMap: nextValueMap }
    }
    default: {
      return state
    }
  }
}

export const useFormState = initialValueMap =>
  useReducer(reducer, getInitialState(initialValueMap))
