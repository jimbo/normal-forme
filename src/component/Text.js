import { createElement, forwardRef, useCallback } from "react"

import { useFormContext } from "../context/form"
import withDefaultProps from "../util/withDefaultProps"

const useDefaultProps = withDefaultProps({
  type: "text",
})

const Text = (props, ref) => {
  const { field, ...restProps } = useDefaultProps(props)
  const { setValue, value } = useFormContext(field)
  const stringValue = value == null ? "" : `${value}`

  const handleChange = useCallback(
    event => {
      setValue(event.target.value)
    },
    [setValue]
  )

  return (
    <input
      {...restProps}
      ref={ref}
      onChange={handleChange}
      value={stringValue}
    />
  )
}

export default forwardRef(Text)
