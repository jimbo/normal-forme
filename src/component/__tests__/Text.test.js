import { createElement } from "react"
import TestRenderer from "react-test-renderer"

import Text from "../Text"
import { FormProvider } from "../../context/form"

const field = "a"
const dispatch = jest.fn()
const getState = value => ({ valueMap: new Map().set(field, value) })

describe("Text", () => {
  it("renders correctly", () => {
    const state = getState("b")
    const tree = TestRenderer.create(
      <FormProvider value={[state, dispatch]}>
        <Text field={field} />
      </FormProvider>
    ).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it("casts `undefined` value to empty string", () => {
    const state = getState()
    const tree = TestRenderer.create(
      <FormProvider value={[state, dispatch]}>
        <Text field={field} />
      </FormProvider>
    ).toJSON()

    expect(tree.props.value).toBe("")
  })

  it("casts `null` value to empty string", () => {
    const state = getState(null)
    const tree = TestRenderer.create(
      <FormProvider value={[state, dispatch]}>
        <Text field={field} />
      </FormProvider>
    ).toJSON()

    expect(tree.props.value).toBe("")
  })

  it("dispatches on change", () => {
    const state = getState("b")
    const nextValue = "c"
    const tree = TestRenderer.create(
      <FormProvider value={[state, dispatch]}>
        <Text field={field} />
      </FormProvider>
    ).toJSON()

    tree.props.onChange({ target: { value: nextValue } })

    expect(dispatch).toHaveBeenCalledTimes(1)
    expect(dispatch).toHaveBeenLastCalledWith({
      payload: { field, nextValue },
      type: "set value",
    })
  })

  it("memoizes its change handler", () => {
    const state = getState("b")
    const nextValue = "c"
    const tree = TestRenderer.create(
      <FormProvider value={[state, dispatch]}>
        <Text field={field} />
      </FormProvider>
    )

    const { onChange: handlerA } = tree.toJSON().props
    const nextState = getState(nextValue)

    tree.update(
      <FormProvider value={[nextState, dispatch]}>
        <Text field={field} />
      </FormProvider>
    )

    const { onChange: handlerB } = tree.toJSON().props

    expect(handlerA).toBeInstanceOf(Function)
    expect(handlerB).toBeInstanceOf(Function)
    expect(handlerB).toBe(handlerA)
  })
})
