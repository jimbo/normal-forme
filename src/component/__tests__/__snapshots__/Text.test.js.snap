// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text renders correctly 1`] = `
<input
  onChange={[Function]}
  type="text"
  value="b"
/>
`;
